### ISSUE TITLE
Chained .with() calls that include nested relationships fail

### ISSUE THREAD

**Issue by clockworked247** [vuex-orm/vuex-orm#102]
Given the following query:

```
Vue.$store.getters['entities/workspaces/query']()
            .with('queries.pane.query').with('queries.pane.explore.obj').with('queries.pane.profile')
            .where('active', true)
            .limit(1).get()[0];
```

Only the last with() call ('queries.pane.profile') has an effect on the item returned from the store.    I've swapped around the with() invocation order and can confirm that only the last relationship is applied.

**Comment by kiaking** [comment]
Just in time, I've wrote this #94 too but you have to nest the queries.

```js
Vue.$store.getters['entities/workspaces/query']()
  .with('queries', (query) => {
    query.with('pane', (query) => {
      query.with('query').with('explore.obj').with('profile')
    })
  })
  .where('active', true)
  .first()
```

I had many people having trouble with this syntax so I'll create another issue to resolve this one too!

**Comment by clockworked247** [comment]
Yea, we crossed paths there... apologies.  Thanks for looking into the fix.  The ability to chain nested relationships would be helpful.  I'll use this method for now.   Thanks!

**Comment by clockworked247** [comment]
I see the dilemma.   Could you only resolve the "has" type (hasMany, hasOne) relationships and ignore the "belongsTo" ones?   This way, you would only be resolving the children/grandchildren/great-grandchildren/etc of the parent query.

Not sure how to make that clear in the API though.   

Maybe I need to better understand how belongsTo() and with() relate.  I hadn't thought of that child-parent relationship.

**Comment by kiaking** [comment]
I wouldn't think adding method that fetch only "has" type method. Because how about when you want to fetch non circular relation with belongsTo? For example when you have something like User hadMany Post, Post belongsTo Project. Wouldn't you want Post to retrieve Project as well?

> Maybe I need to better understand how belongsTo() and with() relate. I hadn't thought of that child-parent relationship.

Well I think your data schema is quite complex. Which I'm very happy because you're using Vuex ORM in such complex project! One thing what I would do is just add helper getters that create complex query and use that when ever you need predefined query scopes.

```js
const getters = {
  withQueryPanes ({ getters }) {
    return getters.query().with('queries', (query) => {
      query.with('pane', (query) => {
        query.withAll().with('explore.obj')
      })
    })
  }
}

// Then use it to fetch predefined with call.
store.getters['entities/workspaces/withQueryPanes']().where('active', true).first()
```

**Comment by clockworked247** [comment]
> One thing what I would do is just add helper getters that create complex query and use that when ever you need predefined query scopes.

@kiaking Very slick.  I like it.  Thanks.

> I wouldn't think adding method that fetch only "has" type method. Because how about when you want to fetch non circular relation with belongsTo?  For example when you have something like User hasMany Post, Post belongsTo Project. Wouldn't you want Post to retrieve Project as well?

Yes, I agree, you would want it to retrieve Project as well.  

Hmm, well my only - likely throwaway - idea would be to then allow for resolution of all relationships up until you get to the same entity type you are querying from and stop there (including any siblings of my same type... ugh, that's where it could get really hairy).

Considering all the relationship types your library supports, that sounds quite challenging to implement.   Might not be worth it.  

Thanks again for your time!

**Comment by kiaking** [comment]
Yes, it does become quite complicated. I think we should introduce some easy way for the user to define their query scopes. Something like [Laravel has](https://laravel.com/docs/5.6/eloquent#query-scopes).

**Comment by timoschwarzer** [comment]
I like the query scopes idea!
For now, I implemented a `withAllRecursive` method that fetches all relations **to a certain depth** (default 3).
See: https://github.com/vuex-orm/vuex-orm/pull/106

**Comment by timoschwarzer** [comment]
See my comment here for the multiple `with` calls: https://github.com/vuex-orm/vuex-orm/issues/104#issuecomment-369725969

**Comment by funtastic-pbc** [comment]
`withAll` is perfect.  Thanks @kiaking
